core/src/main/scala/org/apache/spark/util/CheckpointingIterator.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.io.IOException
+
+import scala.reflect.ClassTag
+
+import org.apache.hadoop.fs.FileSystem
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark._
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.CheckpointRDD
+import org.apache.spark.serializer.SerializationStream
+
+/**
+ * Wrapper around an iterator which writes checkpoint data to HDFS while running action on
+ * a RDD to support checkpointing RDD.
+ */
+private[spark] class CheckpointingIterator[A: ClassTag, +I <: Iterator[A]](
+  sub: I,
+  path: String,
+  broadcastedConf: Broadcast[SerializableConfiguration],
+  partitionId: Int,
+  context: TaskContext,
+  blockSize: Int = -1) extends Iterator[A] with Logging {
+
+  val env = SparkEnv.get
+  var fs: FileSystem = null
+  val bufferSize = env.conf.getInt("spark.buffer.size", 65536)
+  var serializeStream: SerializationStream = null
+
+  var finalOutputPath: Path = null
+  var tempOutputPath: Path = null
+
+  def init(): this.type = {
+    val outputDir = new Path(path)
+    fs = outputDir.getFileSystem(broadcastedConf.value.value)
+
+    val finalOutputName = CheckpointRDD.splitIdToFile(partitionId)
+    finalOutputPath = new Path(outputDir, finalOutputName)
+    tempOutputPath =
+      new Path(outputDir, "." + finalOutputName + "-attempt-" + context.attemptNumber)
+
+    if (fs.exists(tempOutputPath)) {
```

Comment:
It is possible that more than one iterators for the same split are created and used, e.g., `CartesianRDD`. We only need one of them to write checkpoint data to disk.